Description: Use llvm 6.0 on arm*
Author: Gianfranco Costamagna <locutusofborg@debian.org>
Last-Update: 2018-07-19

Index: b/configure.ac
===================================================================
--- a/configure.ac
+++ b/configure.ac
@@ -638,7 +638,7 @@ AC_SUBST([LibtoolCmd])
 # tools we are looking for. In the past, GHC supported a number of
 # versions of LLVM simultaneously, but that stopped working around
 # 3.5/3.6 release of LLVM.
-LlvmVersion=5.0
+LlvmVersion=6.0
 AC_SUBST([LlvmVersion])
 sUPPORTED_LLVM_VERSION=$(echo \($LlvmVersion\) | sed 's/\./,/')
 AC_DEFINE_UNQUOTED([sUPPORTED_LLVM_VERSION], ${sUPPORTED_LLVM_VERSION}, [The supported LLVM version number])
